Major refactor of plugin and support of rules selection

AdrieanKhisbe opened the pull request:
Hello @azeemba :)

First thanks of all for this useful plugin!

Here is roughly what PR consist in:
- update the code to es6 standard
- set up eslint and prettier for the plugin itself
- **add possibility to select rules to check for**.  (see ddad466)
  - plugin is disabled by default, rules have to be enable as error or warning
  - a all rule is available (that can be configure to ignore comments)
  - rule name is `json/the-error-code-in-kebab-case`
- adds integrations tests

Builds of branch can be seen on [travis](https://travis-ci.com/AdrieanKhisbe/eslint-plugin-json/builds), and library is available for test on npm under [`eslint-plugin-json-beta`](https://www.npmjs.com/package/eslint-plugin-json-beta) as `eslint-plugin-json-beta@2.0.0-alpha.1`

Documentation needs to be updated, which I will do after the first feedbacks :)

AdrieanKhisbe committed:
Add some example folder to try out plugin

AdrieanKhisbe committed:
Set up basic eslint for the plugin

AdrieanKhisbe committed:
Refactor and convertion to es6 (arrow&const)

AdrieanKhisbe committed:
Set up and apply prettier

AdrieanKhisbe committed:
Some refactor, powered up by lodash

AdrieanKhisbe committed:
Convert code name to kebab case

AdrieanKhisbe committed:
Major feature to be able to activate or not different rules

AdrieanKhisbe committed:
Handle file comments and bumpt language service

AdrieanKhisbe committed:
Refactor so that comments are in the json/* scope

AdrieanKhisbe committed:
Adapt the tests by adding a fake eslint engine to apply rules

AdrieanKhisbe committed:
Tweak the location and restore source added in the postprocess method

AdrieanKhisbe committed:
Set up integration tests :)

AdrieanKhisbe committed:
Improvement to output of integration test

AdrieanKhisbe committed:
Tweaks to ease handling of a different plugin name

AdrieanKhisbe committed:
Update the .npmignore to be able to publish

AdrieanKhisbe committed:
Allow the customization of the wildcard rule to disable comment errors$

AdrieanKhisbe committed:
Eliminate unused argument from signature of getDiagnotsticss

maybe could go further deleting the potentialy unessary anti duplicate mecanism

AdrieanKhisbe committed:
Refactor the rules and give an alias to '*' (json)

AdrieanKhisbe committed:
Refactor the integration test and add untracked sample

azeemba commented:
@AdrieanKhisbe thanks for the PR!

Due to the size of the PR and some time constraints, it will take me couple of weeks to go through this PR. Hopefully that's okay for you.

If everything looks good, would you be interested in maintaining (or helping maintain) this project as well?

AdrieanKhisbe commented:
Your welcome @azeemba ! :)

No hurry! Take the time needed :)

(And yes, might be interested into giving a hand to main it)

azeemba commented on package.json:
```diff
@@ -22,15 +24,19 @@
+    "eslint": "^5.1.0",
+    "eslint-config-prettier": "^6.2.0",
+    "eslint-plugin-prettier": "^3.1.0",
+    "lodash": "^4.17.15",
```
Looks like lodash is not just a dev dependency right? So we should move it to be a main dependency?

AdrieanKhisbe replied:
hell yeah! Good catch :)

azeemba reviewed (dismissed):
This is awesome, thanks again for your PR!

I have added a suggestion for the integration test and also a question about how "allowComments" is configured right now.

Do you think you can also update the README.md with the information of the rules/config options that are supported?

Since the changes here are significant, I will make a new major release once we land this changeset.

AdrieanKhisbe commented on example/samples/jsonc.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "hello": "world"
```
yep, that's the point. (`jsonc` for json comment, as it's used by [some](https://stackoverflow.com/a/55708553/1806019). 
I can rename it to make it more clear if you want :)

AdrieanKhisbe replied:
indeed a comment was missing  🤦‍♂

AdrieanKhisbe commented on example/.eslintrc:
```diff
@@ -0,0 +1,16 @@
+        {
+            "files": ["samples/jsonc.json"],
+            "rules": {
+                "json/json": ["warn", {"allowComments": true}]
```
It can be specificaly set with `json/comment-not-permitted`.

By default, `json/json` (and it's alias `json/*`), activate this rules (as comments are not part of the json spec). the option `allowComments` allow user to activate everything but this rule.

But it's strange you had to made the change since `json/json` and `json/*` are [supposed to be the same](https://github.com/azeemba/eslint-plugin-json/pull/26/files#diff-6d186b954a58d5bb740f73d84fe39073R121-R122)
I'm going to try to replicate

AdrieanKhisbe commented:
Sure I'll update the README accordingly :)
(And sure, I think also a major is best suited for theses kind of chaneges)

AdrieanKhisbe committed:
Rename example to examples and a a comment in json with comment file

AdrieanKhisbe committed:
Upgrade lodash as production dependency

AdrieanKhisbe committed:
Update all dependencies

AdrieanKhisbe committed:
Update README with new features

AdrieanKhisbe committed:
Minor refactors for more concise code

AdrieanKhisbe committed:
Fix npmignore (examples folder, and add some more tmp/private)

AdrieanKhisbe commented:
@azeemba I updated the documentation.

Tell me if it suits you :)

Here is some things I'd like to do If you agree,
- set up coverage with codecov. (computation, publish, and badge)
- rename `lib` to `src`
- add a `LICENSE` file to the root (it is even mention by the plugin itself https://github.com/azeemba/eslint-plugin-json/blob/master/lib/index.js#L5). 
  MIT maybe, unless you prefer `ISC` that was probably `npm init` default.

azeemba commented:
All those suggestions sound great!
Regarding codecov, we have access to [GitHub Actions](https://github.com/features/actions) so we can switch to using that as well if you are interested. Though probably its best to that on a separate PR.

I think having integration tests like you have provided are super valuable but I think there is some bug in the script at the moment. Currently `json-with-comments` is passing the integration test but it should be failing since `.eslintrc` has not been updated with the updated filename. 

If its okay with you, I would like to add a commit to switch the integration test script to be in JS so the tests can be more structured.

azeemba committed:
Switch integration test script to be in JS

This allows us to do more clear checks on the errors returned from
eslint. The sample projects eslintrc is also updated to "allowComments"

azeemba commented:
Let me know if you have any concerns with the switch in the integration tests!

AdrieanKhisbe commented:
No problem @azeemba ! :)
Good work with `a35309b``

Will then do the suggestions you accepted, 
and possibly split the real integration test you created to have the _engine_, and just the test in a dedicated file [with this](https://github.com/azeemba/eslint-plugin-json/pull/26/commits/a35309bbea01142514f4f662c3d8f6ca4fcd2d96#diff-92a99879d26f71cc020e690abbfbbab8R74-R82)

Will have a look of github actions. Sounds promising but haven't try it for now. This could be the opportunity :)

AdrieanKhisbe committed:
Move source from lib to src

AdrieanKhisbe committed:
Rework integrations tests :hammer_and_pick:

AdrieanKhisbe committed:
Improvements to plugin making it resilient to unfound lint results :shield:

AdrieanKhisbe committed:
Set up coverage with nyc :dart:

AdrieanKhisbe committed:
Single module.exports :open_file_folder:

AdrieanKhisbe committed:
Set up the MIT License :balance_scale:

AdrieanKhisbe commented:
Hey @azeemba :)
I just implemented the suggestions I made last time.

We might be good to go :)

azeemba reviewed (approved):
Thanks for all your contributions!

Since you have merge privileges now, I will let you merge the PR. I will push out a new major release after that!

AdrieanKhisbe commented:
And here it is then :)
